File: java/110.balanced-binary-tree.java
/*
 * @lc app=leetcode id=110 lang=java
 *
 * [110] Balanced Binary Tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    boolean bal = true;

    public boolean isBalanced(TreeNode root) {
        height(root);
        return bal;
    }

    private int height(TreeNode root) {
        if (!bal) return 0;
        if (root == null) return -1;
        int lh = height(root.left);
        int rh = height(root.right);
        if (Math.abs(lh - rh) > 1) bal = false;
        return 1 + Math.max(lh, rh);
    }
}

File: go/110.balanced-binary-tree.go
/*
 * @lc app=leetcode id=110 lang=golang
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	_, bal := dfsCheck(root)
	return bal
}

func dfsCheck(node *TreeNode) (height int, bal bool) {
	if node != nil {
		lh, lb := dfsCheck(node.Left)
		if !lb {
			return
		}
		rh, rb := dfsCheck(node.Right)
		if !rb || lh - rh < -1 || lh - rh > 1 {
			return
		}
		if lh < rh {
			height = 1 + rh
		} else {
			height = 1 + lh
		}
	}
	bal = true
	return
}

// @lc code=end

